### DISCUSSION TITLE
Middleware async setup function

### DISCUSSION THREAD

**Discussion by ottermata** [litestar-org/litestar#855] [Q&A] [answered]
Hey, is there a way to have an async setup function for a middleware?

Currently I'm just running it within the `__call__` method, if it hasn't been executed before, but that doesn't seem ideal.

**Comment by provinzkraut** [comment, marked as answer]
Can you elaborate a bit / provide an example? What exactly do you mean by "setup function"?

**Reply by ottermata** [reply]
As part of an authentication middleware I need the JWKs from KeyCloak so that I can validate the tokens.
The setup would be a downloading the JWKs.

**Reply by provinzkraut** [reply]
So this is something that needs to happen only once? I'm still not entirely should what you're trying to do here. Could you maybe provide some code, e.g. the current solution you're using?

If this needs to happen only once, I'd do it at app startup.

**Reply by ottermata** [reply]
So an on_startup handler function that stores the result in the app state? I guess that's better.

This is roughly what I was using:
```python
class AuthMiddleware(AbstractAuthenticationMiddleware):
    # set _jwks to None in init
    async def _get_jwk(self, connection: ASGIConnection, key_id: str) -> JWKData | None:
        if self._jwks is None:
            self._jwks = await self._fetch_jwks(connection)

        return self._jwks.get(key_id)

    async def authenticate_request(self, connection: ASGIConnection) -> AuthenticationResult:
        token = _get_bearer_token(connection)
        key_id = _get_key_id(token)
        if key := await self._get_jwk(connection, key_id):
            decoded = key.decode(token)
            return AuthenticationResult(...)

        raise NotAuthorizedException()
```

**Reply by provinzkraut** [reply]
> So an on_startup handler function that stores the result in the app state? I guess that's better.

Yes, something like that.
